File: src/BjjEire.Seeder/SeederService.cs
using BjjEire.Domain.Entities;

namespace BjjEire.Seeder;

public class SeederService(IMongoDatabase db, bool dryRun)
{
    private enum UpsertOutcome { Inserted, Replaced, DryRunInsert, DryRunReplace, Failed }


    public async Task<int> SeedAsync<T>(string collectionName, string jsonPath) where T : BaseEntity
    {
        await Console.Out.WriteLineAsync($"\n── {collectionName} ({jsonPath}) ──");

        List<T>? entities = await LoadEntitiesAsync<T>(jsonPath);
        if (entities is null)
            return 1;
        if (entities.Count == 0)
        {
            await Console.Out.WriteLineAsync("  No documents found — skipping.");
            return 0;
        }

        IMongoCollection<T> collection = db.GetCollection<T>(collectionName);
        (int inserted, int replaced, int failed) counts = await UpsertAllAsync(collection, entities);

        await Console.Out.WriteLineAsync($"  Result: {counts.inserted} inserted, {counts.replaced} replaced, {counts.failed} failed.");
        return counts.failed > 0 ? 1 : 0;
    }

    private static async Task<List<T>?> LoadEntitiesAsync<T>(string jsonPath) where T : BaseEntity
    {
        (List<T>? entities, string? error) = await EntityLoader.LoadAsync<T>(jsonPath, EntityLoader.PermissiveOptions);
        if (error is not null)
            await Console.Error.WriteLineAsync($"  ERROR: {error}");
        return entities;
    }

    private async Task<(int inserted, int replaced, int failed)> UpsertAllAsync<T>(
        IMongoCollection<T> collection, List<T> entities) where T : BaseEntity
    {
        int inserted = 0;
        int replaced = 0;
        int failed = 0;

        foreach (T entity in entities)
        {
            (UpsertOutcome outcome, string? error) = await UpsertOneAsync(collection, entity);
            LogOutcome(entity.Id, outcome, error);

            switch (outcome)
            {
                case UpsertOutcome.Inserted:
                    inserted++;
                    break;
                case UpsertOutcome.Replaced:
                    replaced++;
                    break;
                case UpsertOutcome.Failed:
                    failed++;
                    break;
            }
        }

        return (inserted, replaced, failed);
    }

    private async Task<(UpsertOutcome outcome, string? error)> UpsertOneAsync<T>(
        IMongoCollection<T> collection, T entity) where T : BaseEntity
    {
        FilterDefinition<T> filter = Builders<T>.Filter.Eq(e => e.Id, entity.Id);

        try
        {
            if (dryRun)
            {
                bool exists = await collection.CountDocumentsAsync(filter) > 0;
                return (exists ? UpsertOutcome.DryRunReplace : UpsertOutcome.DryRunInsert, null);
            }

            ReplaceOneResult result = await collection.ReplaceOneAsync(filter, entity, new ReplaceOptions { IsUpsert = true });
            return result.UpsertedId is not null
                ? (UpsertOutcome.Inserted, null)
                : (UpsertOutcome.Replaced, null);
        }
        catch (Exception ex)
        {
            return (UpsertOutcome.Failed, ex.Message);
        }
    }

    private static void LogOutcome(string id, UpsertOutcome outcome, string? error) =>
        Console.WriteLine(outcome switch
        {
            UpsertOutcome.Inserted => $"  INSERTED      {id}",
            UpsertOutcome.Replaced => $"  REPLACED      {id}",
            UpsertOutcome.DryRunInsert => $"  [DRY RUN]     {id} — would insert",
            UpsertOutcome.DryRunReplace => $"  [DRY RUN]     {id} — would replace",
            UpsertOutcome.Failed => $"  FAILED        {id} — {error}",
            _ => $"  UNKNOWN       {id}",
        });
}

File: src/bjjeire-api/src/main/java/com/bjjeire/api/seeder/SeederService.java
package com.bjjeire.api.seeder;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.mongodb.bulk.BulkWriteResult;
import java.io.IOException;
import java.nio.file.Path;
import java.util.Collection;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import org.springframework.data.mongodb.core.BulkOperations;
import org.springframework.data.mongodb.core.MongoTemplate;
import org.springframework.data.mongodb.core.query.Criteria;
import org.springframework.data.mongodb.core.query.Query;
import org.springframework.data.util.Pair;

public class SeederService {
    public record SeedResult(int exitCode, List<String> ids) {}

    private final MongoTemplate mongoTemplate;
    private final boolean dryRun;

    public SeederService(MongoTemplate mongoTemplate, boolean dryRun) {
        this.mongoTemplate = mongoTemplate;
        this.dryRun = dryRun;
    }

    public SeedResult seed(SeederCollection collection, Path jsonPath, ObjectMapper mapper) {
        System.out.printf("%n── %s (%s) ──%n", collection.name(), jsonPath);

        List<Object> entities;
        try {
            entities = mapper.readValue(
                    jsonPath.toFile(), mapper.getTypeFactory().constructCollectionType(List.class, collection.type()));
        } catch (IOException exception) {
            System.err.println("  ERROR: " + exception.getMessage());
            return new SeedResult(ExitCodes.FAILURE, List.of());
        }

        if (entities == null || entities.isEmpty()) {
            System.out.println("  No documents found — skipping.");
            return new SeedResult(ExitCodes.SUCCESS, List.of());
        }

        List<String> ids = entities.stream().map(collection.idOf()).toList();

        if (dryRun) {
            return dryRunReport(collection, ids);
        }

        return bulkUpsert(collection, entities, ids);
    }

    public int prune(SeederCollection collection, Collection<String> keepIds) {
        if (keepIds.isEmpty()) {
            System.out.printf("  Skipping prune for %s: no seeded ids (empty source).%n", collection.name());
            return ExitCodes.SUCCESS;
        }

        Query staleQuery = new Query(Criteria.where("_id").nin(keepIds));

        if (dryRun) {
            long stale = mongoTemplate.count(staleQuery, collection.type());
            System.out.printf("  [DRY RUN] Would prune %d stale document(s) from %s.%n", stale, collection.name());
            return ExitCodes.SUCCESS;
        }

        long deleted = mongoTemplate.remove(staleQuery, collection.type()).getDeletedCount();
        System.out.printf("  Pruned %d stale document(s) from %s.%n", deleted, collection.name());
        return ExitCodes.SUCCESS;
    }

    private SeedResult dryRunReport(SeederCollection collection, List<String> ids) {
        Query existsQuery = new Query(Criteria.where("_id").in(ids));
        existsQuery.fields().include("_id");
        Set<String> existing = new HashSet<>(mongoTemplate.find(existsQuery, collection.type()).stream()
                .map(collection.idOf())
                .toList());
        long replace = ids.stream().filter(existing::contains).count();
        System.out.printf("  [DRY RUN] Would insert %d, replace %d.%n", ids.size() - replace, replace);
        return new SeedResult(ExitCodes.SUCCESS, ids);
    }

    private SeedResult bulkUpsert(SeederCollection collection, List<Object> entities, List<String> ids) {
        // Seeder writes bypass the service layer, so the TTL stamp is applied here,
        // matching the expiry-stamping behaviour of the API's write path.
        for (Object entity : entities) {
            collection.stampExpiry().accept(entity);
        }

        BulkOperations bulk = mongoTemplate.bulkOps(BulkOperations.BulkMode.UNORDERED, collection.type());
        for (Pair<String, Object> entry : zip(ids, entities)) {
            bulk.replaceOne(
                    new Query(Criteria.where("_id").is(entry.getFirst())),
                    entry.getSecond(),
                    org.springframework.data.mongodb.core.FindAndReplaceOptions.options()
                            .upsert());
        }

        try {
            BulkWriteResult result = bulk.execute();
            System.out.printf(
                    "  Result: %d inserted, %d replaced.%n", result.getUpserts().size(), result.getModifiedCount());
            return new SeedResult(ExitCodes.SUCCESS, ids);
        } catch (RuntimeException exception) {
            System.err.println("  FAILED: " + exception.getMessage());
            return new SeedResult(ExitCodes.FAILURE, ids);
        }
    }

    private static List<Pair<String, Object>> zip(List<String> ids, List<Object> entities) {
        return java.util.stream.IntStream.range(0, ids.size())
                .mapToObj(i -> Pair.of(ids.get(i), entities.get(i)))
                .toList();
    }
}
